discovery/sync_manager.go

```diff
@@ -0,0 +1,651 @@
+	// active syncer in the queue to ensure it is not stale.
+	pendingActiveSyncers map[routing.Vertex]*activeSyncer
+
+	wg   sync.WaitGroup
+	quit chan struct{}
+}
+
+// newSyncManager constructs a new SyncManager backed by the given config.
+func newSyncManager(cfg *SyncManagerCfg) *SyncManager {
+	return &SyncManager{
+		cfg: *cfg,
+		activeSyncers: make(
+			map[routing.Vertex]*GossipSyncer, cfg.NumActiveSyncers,
+		),
+		inactiveSyncers:          make(map[routing.Vertex]*GossipSyncer),
+		pendingActiveSyncerQueue: list.New(),
+		pendingActiveSyncers:     make(map[routing.Vertex]*activeSyncer),
+		newActiveSyncers:         make(chan *activeSyncer),
+		staleActiveSyncers:       make(chan routing.Vertex),
+		quit:                     make(chan struct{}),
+	}
+}
+
+// Start starts the SyncManager in order to properly carry out its duties.
+func (m *SyncManager) Start() {
+	m.start.Do(func() {
+		m.wg.Add(1)
+		go m.syncerHandler()
+	})
+}
+
+// Stop stops the SyncManager from performing its duties.
+func (m *SyncManager) Stop() {
+	m.stop.Do(func() {
+		close(m.quit)
+		m.wg.Wait()
+
+		m.syncersMtx.Lock()
+		defer m.syncersMtx.Unlock()
+
+		for _, syncer := range m.activeSyncers {
+			syncer.Stop()
+		}
+		for _, syncer := range m.inactiveSyncers {
+			syncer.Stop()
+		}
+	})
+}
+
+// enqueueActiveSyncer adds the given gossip syncer to the end of the
+// round-robin queue.
+func (m *SyncManager) enqueueActiveSyncer(s *GossipSyncer, transition bool) {
+	select {
+	case m.newActiveSyncers <- &activeSyncer{
+		GossipSyncer: s,
+		transition:   transition,
+	}:
+	case <-m.quit:
+	}
+}
+
+// dequeueActiveSyncer removes the syncer for the given peer from the
+// round-robin queue.
+func (m *SyncManager) dequeueActiveSyncer(peer routing.Vertex) {
+	select {
+	case m.staleActiveSyncers <- peer:
+	case <-m.quit:
+	}
+}
+
+// syncerHandler is the SyncManager's main event loop responsible for:
+//
+//   1. Finding new peers to receive graph updates from to ensure we don't only
+//   receive them from the same set of peers.
+//
+//   2. Finding new peers to force a full sync with to ensure we have as much of
+//   the public network as possible.
+//
+//   3. Managing the round-robin queue of our active syncers to ensure they
+//   don't overlap and request the same set of channels, which significantly
+//   reduces bandwidth usage.
+//
+// NOTE: This must be run as a goroutine.
+func (m *SyncManager) syncerHandler() {
+	defer m.wg.Done()
+
+	// We'll start all of our relevant tickers, except for our
+	// ActiveSyncerTimeoutTicker, which should only be started when we start
+	// an activeSyncer.
+	m.cfg.RotateTicker.Resume()
+	defer m.cfg.RotateTicker.Stop()
+
+	m.cfg.HistoricalSyncTicker.Resume()
+	defer m.cfg.HistoricalSyncTicker.Stop()
+
+	defer m.cfg.ActiveSyncerTimeoutTicker.Stop()
+
+	var (
+		// current will hold the current activeSyncer we're waiting for
+		// to complete its state machine.
+		current *activeSyncer
```

Comment:
Doesn't this contradict the fact that we can have multiple active syncers?

Reply:
I think there's some confusion about what the round-robin is meant to achieve. See my previous comment: https://github.com/lightningnetwork/lnd/pull/2740#discussion_r270156880